Davem/rc3

iabyn opened the pull request:
Third set of PERL_RC_STACK commits. This PR mainly includes lots of unwrapping of list ops. It also includes a couple of new API functions:
rpp_replace_at() and rpp_context().

There are still a few 'hot' list functions still to be unwrapped, such as pp_multiconcat(), but currently 10 runs of the test suite take about 5% more CPU on average with -DPERL_RC_STACK.

iabyn committed:
make RC-stack-aware: unwrap startmatch and misc

Remove the temporary wrappers from these pp() functions:

    pp_smartmatch
    pp_i_eq
    pp_eq
    pp_seq
    pp_match

They had to be treated as one group since pp_smartmatch() tail calls
the other four pp() functions.

iabyn committed:
make RC-stack-aware: unwrap pp_padsv_store()

Remove the temporary wrapper from this unary pp() function.
It should have been done at the same time as all the other hot unary
ops, but I missed it.

iabyn committed:
make RC-stack-aware: unwrap pp_join()

Remove the temporary wrapper from pp_join()

iabyn committed:
make RC-stack-aware: unwrap pp_print(),tied_method

Remove the temporary wrapper from pp_print().

Also, update Perl_tied_method() to support both ref-counted-stack and
non-ref-counted-stack callers. Initially pp_print() is the only RC
caller of it.

Perl_tied_method() was a bit of mess: undocumented and confusing.
So also add some more code comments, and rename one of its parameters
from 'sp' to 'mark' to better reflect that it points to the base of the
argument list, not to the top.

iabyn committed:
make RC-stack-aware: unwrap pp_push()

Remove the temporary wrapper from pp_push()

iabyn committed:
make RC-stack-aware: unwrap pp_list()

Remove the temporary wrapper from pp_list()

iabyn committed:
make RC-stack-aware: unwrap pp_unshift()

Remove the temporary wrapper from pp_unshift()

iabyn committed:
make RC-stack-aware: unwrap pp_refgen()

Remove the temporary wrapper from pp_refgen()

iabyn committed:
make RC-stack-aware: pp_anonlist, pp_av2arylen

Remove the temporary wrapper from pp_anonlist() and update it to directly
work in a refcounted-stack environment.

In addition, fix a bug in (the already unwrapped) pp_av2arylen(),
which only manifests itself in PERL_RC_STACK environments.

In particular, this expression in lvalue context was triggering a
premature free of the array:

    [1..10]->$#*

Basically pp_av2arylen() pops the AV off the stack and replaces it
with an SV with PERL_MAGIC_arylen which points back to the AV (and vice
versa). The link back is weak, so nothing actually keeps the AV alive;
previously, something else like the temps stack was keeping AV alive.

So swap the weak/strong mg_ptrs around to make the problem go away.

iabyn committed:
make RC-stack-aware: unwrap pp_anonhash()

Remove the temporary wrapper from pp_anonhash()

iabyn committed:
make RC-stack-aware: unwrap: pp_(kv)[ah]slice

Remove the temporary wrappers from:
    pp_aslice()
    pp_hslice()
    pp_kvaslice()
    pp_kvhslice()

iabyn committed:
add rpp_replace_at() API function

iabyn committed:
add rpp_context() API function

rwp0 commented:
Sorry but:

What is PERL_RC_STACK define (build option)?

Is there docs to read about this?

Thanks

iabyn committed:
make RC-stack-aware: unwrap pp_multideref()

Remove the temporary wrapper from pp_multideref()

iabyn committed:
make RC-stack-aware: unwrap pp_multiconcat()

Remove the temporary wrapper from pp_multiconcat()

iabyn committed:
threads->object(): handle no argument

threads->object() without an argument is supposed to return undef.
But before checking that there's no argument, the XS was actually
accessing the argument.

This went undetected for around 13 years, due to the fact that
1) likely nothing outside the test code  in basic.t called it without an
   argument;
2) as long as any old SV pointer was still on the stack it would
(probably) harmlessly access it and then return.

I only spotted it due to the fact that a recent change to PERL_RC_STACK
meant that the pointer one above the top of stack happened to be NULL.

richardleach commented:
> currently 10 runs of the test suite take about 5% more CPU on average with -DPERL_RC_STACK.

Only about 5% at this stage seems great: a class of bugs dealt with, not all ops yet unwrapped, and unlocked possibilities for creating fewer temporaries (therefore a more efficient runtime overall).

jkeenan commented:
[snip]
> 
> There are still a few 'hot' list functions still to be unwrapped, such as pp_multiconcat(), but currently 10 runs of the test suite take about 5% more CPU on average with -DPERL_RC_STACK.

How are you measuring CPU?  (I've been focusing on running time, which looks good.)  thanks.

iabyn commented:
On Sat, Sep 09, 2023 at 01:30:46PM -0700, James E Keenan wrote:
> How are you measuring CPU?  (I've been focusing on running time, which looks good.)  thanks.

Very crude at the moment. Just running the test suite 10 times and taking
the average of the csys value from the final
"Files=2776, .... csys = 1809.66 CPU)"
line.

It's more an indication at this point that the slowdown is nearer 5% than
50% say.

I'll start benchmarking more accurately once all the hot ops are
unwrapped.

-- 
Decaffeinated coffee is like dehydrated water

tonycoz commented on inline.h:
```diff
@@ -705,6 +705,45 @@ Perl_rpp_replace_at(pTHX_ SV **sp, SV *sv)
+Perl_rpp_context(pTHX_ SV **mark, U8 gimme, SSize_t extra)
+{
+    PERL_ARGS_ASSERT_RPP_CONTEXT;
+
```
Would it be worth asserting that `mark <= PL_stack_sp - extra` ?

Possibly also `extra >= 0`.

iabyn commented:
On Sun, Sep 10, 2023 at 08:32:32PM -0700, Tony Cook wrote:
> Would it be worth asserting that `mark <= PL_stack_sp - extra` ?
> Possibly also `extra >= 0`.

Done, rebased and pushed. I also included 3 more commits:

    threads->object(): handle no argument
    make RC-stack-aware: unwrap pp_multideref()
    make RC-stack-aware: unwrap pp_multiconcat()


-- 
"You may not work around any technical limitations in the software"
    -- Windows Vista license

tonycoz commented:
> make RC-stack-aware: unwrap pp_multideref()
> make RC-stack-aware: unwrap pp_multiconcat()

This turned out to be a lot simpler than I dreaded.

iabyn commented:
On Mon, Sep 11, 2023 at 10:25:41PM -0700, Tony Cook wrote:
> > make RC-stack-aware: unwrap pp_multideref()
> > make RC-stack-aware: unwrap pp_multiconcat()
> 
> This turned out to be a lot simpler than I dreaded.

Yeah that's why I left them till later, then was pleasantly surprised!

However, I'm currently working pp_aasign, and that's even worse than I
was dreading!

-- 
My get-up-and-go just got up and went.
